google/cloud/forseti/scanner/audit/log_sink_rules_engine.py

```diff
@@ -0,0 +1,488 @@
+            self._load_rule_definitions(),
+            snapshot_timestamp=self.snapshot_timestamp)
+
+    def find_violations(self, resource, log_sinks, force_rebuild=False):
+        """Determine whether a resources's log sink config violates rules.
+
+        Args:
+            resource (gcp_type): The resource that the log sinks belong to.
+            log_sinks (list): list of LogSinks for resource.
+            force_rebuild (bool): If True, rebuilds the rule book.
+                This will reload the rules definition file and add the
+                rules to the book.
+
+        Returns:
+            iterable: A generator of rule violations.
+        """
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+
+        violations = self.rule_book.find_violations(resource, log_sinks)
+
+        return set(violations)
+
+    def add_rules(self, rules):
+        """Add rules to the rule book.
+
+        Args:
+            rules (list): The list of rules to add to the book.
+        """
+        if self.rule_book is not None:
+            self.rule_book.add_rules(rules)
+
+
+def _parse_sink_rule_spec(sink_spec):
+    """Validates and escapes a sink from a rule config.
+
+    Args:
+        sink_spec (dict): A sink definition from a LogSink rule definition.
+
+    Returns:
+        dict: A sink definition with fields escaped and globified, or None if
+        sink_spec is invalid.
+    """
+    if not sink_spec:
+        return None
+
+    sink_destination = sink_spec.get('destination')
+    sink_filter = sink_spec.get('filter')
+    sink_include_children = sink_spec.get('include_children')
+    # All fields are mandatory.
+    if any(item is None for item in [
+            sink_destination, sink_filter, sink_include_children]):
+        return None
+
+    # include_children will either match a boolean, or allow either.
+    if sink_include_children.lower() not in ['*', 'true', 'false']:
+        return None
+    if sink_include_children != '*':
+        sink_include_children = sink_include_children.lower() == 'true'
+    return {
+        'destination': escape_and_globify(sink_destination),
+        'filter': escape_and_globify(sink_filter),
+        'include_children': sink_include_children,
+    }
+
+
+class LogSinkRuleBook(bre.BaseRuleBook):
+    """The RuleBook for Log Sink configs.
+
+    Rules from the rules definition file are parsed and placed into a map, which
+    associates the applies_to value and GCP resource (project, folder or
+    organization) with the rules defined for it.
+
+    Resources are evaulated against matching rules defined with applies_to =
+    "self". Project resources are also evaulated against rules for ancestor
+    resources defined with applies_to = "children".
+    """
+
+    supported_resource_types = frozenset([
+        'project',
+        'folder',
+        'organization',
+    ])
+
+    supported_rule_applies_to = frozenset([
+        'self',
+        'children',
```

Comment:
I couldn't think of a case, since the log sink definitions are probably going to be specific to the resource type. Maybe for a blacklist that stops all log sinks, but that can also be done by having both a 'self' and a 'children' rule.